### PR TITLE
Parse missing XPath-related features

### PR DESCRIPTION
**Pull request by DrRataplan**
Throw errors for now, but make the errors 'not implemented' errors instead of parsing errors.

TODO: Tests

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -348,6 +350,35 @@ export function generateParser(options: { language: string }): (input: string) =
 		['NormalChar']
 	);
 
+	// BackReference (XPath only)
+	const backReference: Parser<any> = map(
```
Comment (DrRataplan) [inline src/parser.ts]: TODO: Parser<number>

**Comment by DrRataplan** [comment]
I added the tests. @bwrrp could you give this a look?

File: src/parser.ts [hunk 2]
```diff
@@ -348,6 +350,35 @@ export function generateParser(options: { language: string }): (input: string) =
...
+				star(
+					map(
+						filter(codepoint, charRangePredicate(ZERO_CODE_POINT, asCodepoint('9')), [
+							'digit'
+						]),
+						codepoint => codepoint - ZERO_CODE_POINT
+					)
+				),
+				(firstDigit, restDigits) => {
+					restDigits.reduce((sum, digit) => sum * 10 + digit, firstDigit);
```
Comment (bwrrp) [inline src/parser.ts:373]: Creative! As an alternative, you could use the `recognize` combinator (and isDigit from below), with something like:
```
const backReferenceNumber: Parser<number> = map(
    recognize(
        preceded(
            filter(codepoint, charRangePredicate(asCodepoint('1'), asCodepoint('9')), ['digit']),
            star(filter(codepoint, isDigit, ['digit']))
        )
    ),
    str => parseInt(str, 10)
);
```

File: src/sets.ts [hunk 3]
```diff
@@ -1,19 +1,28 @@
-import { names as blockNames, lengths as blockLengths } from './generated/blocks.json';
+import {
+	charRange,
+	everything,
+	INPUT_END_SENTINAL,
```
Comment (bwrrp) [inline src/sets.ts]: spelling nit: sentinel
(same for other occurrences below)

File: test/xspattern.tests.ts [hunk 4]
```diff
@@ -27,7 +27,7 @@ function check(pattern: string, examples: string[], counterExamples: string[] =
 	});
 }
 
-describe('xspattern', () => {
+describe.only('xspattern', () => {
```
Comment (bwrrp) [inline test/xspattern.tests.ts]: leftover from testing?

File: test/xspattern.tests.ts [hunk 5]
```diff
@@ -197,6 +233,17 @@ describe('xspattern', () => {
 		check('[abcd-[cdef]]', ['a', 'b'], ['c', 'd', 'e', 'f', 'g']);
 		check('[a-z-[aeoui]]', ['b', 'd', 'z'], ['a', 'u', 'o']);
 		check('[a--[a]]', ['-'], ['a', 'b']);
+		check('a[^b]', ['a ', 'ax', 'abaxb'], ['a', 'ab', 'abc'], true);
+	});
+
+	it('throws the correct error when it sees backreferences', () => {
```
Comment (bwrrp) [inline test/xspattern.tests.ts:239]: Could you also add a few tests that show that XPath-specific syntax such as backreferences throws appropriate parsing errors (or may be interpreted as a different type of expression) in XSD mode?

**Review by bwrrp — CHANGES_REQUESTED** [review]
This is great, thanks! I've left a few comments on minor nits.

**Pull request merged**